[Cocoa] Support ST-2094-40 HDR metadata in WebM

jernoble opened the pull request:
#### 21cf922c1e0eb63750d31d1ee867adde01b7a687
<pre>
[Cocoa] Support ST-2094-40 HDR metadata in WebM
<a href="https://bugs.webkit.org/show_bug.cgi?id=286632">https://bugs.webkit.org/show_bug.cgi?id=286632</a>
<a href="https://rdar.apple.com/143771446">rdar://143771446</a>

Reviewed by Jean-Yves Avenard.

Add support for parsing HDR10+ (aka SMPTE ST 2094-40) per-frame metadata
from WebM&apos;s BlockAdditional element, and passing that metadata through
the resulting CMSampleBuffer.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::toCMSampleBuffer):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnBlockGroupBegin):
(WebCore::WebMParser::OnBlockGroupEnd):
(WebCore::WebMParser::VideoTrackData::consumeAdditionalBlockData):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::TrackData::consumeAdditionalBlockData):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/290180@main">https://commits.webkit.org/290180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbea4d89cb9c7788a74ebda1b55919862ced07bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39605 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16554 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68471 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26155 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91848 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6688 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48836 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35223 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38713 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76807 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36204 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95655 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16026 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11715 "Found 1 new test failure: imported/w3c/web-platform-tests/css/selectors/invalidation/modal-pseudo-class-in-has.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77341 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16282 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76634 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76625 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18964 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21013 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19818 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9091 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16040 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21351 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15781 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19232 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/MediaSample.h:
```diff
@@ -218,6 +219,8 @@ class MediaSamplesBlock {
         MediaTime duration { MediaTime::zeroTime() };
         std::pair<MediaTime, MediaTime> trimInterval { MediaTime::zeroTime(), MediaTime::zeroTime() };
         MediaSampleDataType data;
+        MediaSampleDataType hdrMetadata { nullptr };
```
We may want this to be a SharedBuffer instead of a FragmentedSharedBuffer and make it contiguous upon demuxing considering it's pretty small to start.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -316,6 +316,10 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+
+            // Attach HDR10+ (aka SMPTE ST 2094-40) metadata, if present:
+            if (samples[i].hdrMetadataType == HdrMetadataType::SmpteSt209440 && samples[i].hdrMetadata)
+                CFDictionarySetValue(attachments, PAL::kCMSampleAttachmentKey_HDR10PlusPerFrameData, samples[i].hdrMetadata->makeContiguous()->createCFData().get());
```
the makeContiguous is unsightly :)

jyavenard commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1099,6 +1118,37 @@ WebMParser::ConsumeFrameDataResult WebMParser::VideoTrackData::consumeFrameData(
+        if (!std::ranges::equal(s_ST2094_40_Message, blockMoreDataStart))
+            continue;
+
+        auto hdrMetadata = FragmentedSharedBuffer::create(std::span { blockMore.value().data.value() });
```
So here we should create a SharedBuffer instead.

jyavenard replied:
```suggestion
        Ref hdrMetadata = SharedBuffer::create(std::span { blockMore.value().data.value() });
```

jyavenard commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1099,6 +1118,37 @@ WebMParser::ConsumeFrameDataResult WebMParser::VideoTrackData::consumeFrameData(
+
+        auto hdrMetadata = FragmentedSharedBuffer::create(std::span { blockMore.value().data.value() });
+        for (auto& sample : m_pendingMediaSamples) {
+            sample.hdrMetadata = hdrMetadata.ptr();
```
If we replaced the `Ref` above (or `auto` as written) we don't have to pass the ptr().

jyavenard commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -8866,6 +8866,8 @@ header: <WebCore/MediaSample.h>
     MediaTime duration;
     std::pair<MediaTime, MediaTime> trimInterval;
     RefPtr<WebCore::FragmentedSharedBuffer> data;
+    RefPtr<WebCore::FragmentedSharedBuffer> hdrMetadata;
```
```suggestion
    RefPtr<WebCore::SharedBuffer> hdrMetadata;
```

jyavenard commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -919,23 +919,42 @@ webm::Status WebMParser::OnBlockGroupBegin(const webm::ElementMetadata&, webm::A
 
+    // Flush any pending samples, so that the only pending samples after BlockGroup ends
+    // are those from this block group.
+    flushPendingVideoSamples();
```
The `VideoTrackData::flushPendingSamples()` only keeps frames that have the same timestamp and to determine the duration of the frame.
At the most in practice it will have only 2 frames. 

Can we detect that this file will have HDR metadata? if so we may have to delay flushing the pending video frames until group block end is seen. 

If we do that however it will also break the search for a keyframe. We'll need to split up frames with different time, so we need to make sure the resulting CMSampleBuffer only ever contains frame that have the same timestamp

jernoble replied:
A BlockGroup can only have a single Block (and therefore Frame) in it. So this codepath benefits from the buffering you added for timestamp and duration resolution and does not add any additional buffering. I don't believe we need to add an explicit "don't flush if we're in a group" logic, because of the "only one Block in a BlockGroup" rule.

jernoble committed:
[Cocoa] Support ST-2094-40 HDR metadata in WebM
https://bugs.webkit.org/show_bug.cgi?id=286632
rdar://143771446

Reviewed by Jean-Yves Avenard.

Add support for parsing HDR10+ (aka SMPTE ST 2094-40) per-frame metadata
from WebM's BlockAdditional element, and passing that metadata through
the resulting CMSampleBuffer.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::toCMSampleBuffer):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnBlockGroupBegin):
(WebCore::WebMParser::OnBlockGroupEnd):
(WebCore::WebMParser::VideoTrackData::consumeAdditionalBlockData):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::TrackData::consumeAdditionalBlockData):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/290180@main
